### PR TITLE
feat: switched dataloader to grain

### PR DESCRIPTION
**Pull request by emergenz**
- switched dataloader to grain
- dataloader is now fully independent of `tenorflow`/`tf.data`
- switched dataset format to [array_records](https://github.com/google/array_record), which is optimized for parallel random access
- to be fully compute-bound when the dataset is on GPFS, we have to split the videos into 16s (160 frame) chunks (like genie does), and have ~100 records per array_record
  - added utilities for this & conversion from `tfrecord` to `array_record`
- tested on knoms

### CODE REVIEW

**Comment by emergenz** [comment]
- switched to checkpoint manager
- implemented (multi-process) dataloader state checkpointing + restore

**Comment by emergenz** [comment]
### API change in training scripts
- Removed `--checkpoint`
- Added `--save_ckpt` and `--restore_ckpt`

The checkpoint manager is now always initialized with the `--ckpt_dir`, but only restores/saves if the corresponding flag is set. Most importantly, we don't have separate saving/ restoration directories, but the same directory is used.

**Review by maharajamihir — APPROVED** [review]
i tested it and it works well. I could reproduce the training curves of the runs with the old dataloader.
LAM training is slower now (likely IO bound) but tokenizer and dynamics are just as fast as with tf_records.
The shuffling does not eat my RAM anymore and seems like we have perfect shuffling. So that is great.

## tldr: lgtm
